File: topoflow/governance.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Discover Legal
"""Governance (spec §9 Layer 5): pre-flight, halting, policy violations."""
from __future__ import annotations

from .config import Config


class GovernanceAbort(RuntimeError):
    """Raised by preflight for unrecoverable conditions (bad key/quota/etc.)."""


def preflight(ctx, cfg: Config, transport) -> None:
    """Layer 5 pre-flight: abort early on unrecoverable transport problems."""
    if transport is None:
        raise GovernanceAbort("no transport configured")
    if not ctx.prompt:
        raise GovernanceAbort("empty task prompt")


def violation(trace, cfg: Config) -> bool:
    """Detect a policy violation that should halt the trajectory.

    Conservative guards that protect against runaway loops / repeated failures.
    """
    if trace.retries > cfg.max_retries:
        return True
    return False

File: biglaw-go/internal/topoflow/governance.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 Discover Legal

package topoflow

import "errors"

// ErrGovernanceAbort is returned by preflight for unrecoverable conditions.
var ErrGovernanceAbort = errors.New("governance abort")

// preflight is the Layer-5 pre-flight check (spec §9).
func preflight(ctx TaskContext, cfg Config, tx Transport) error {
	if tx == nil {
		return errGov("no transport configured")
	}
	if ctx.Prompt == "" {
		return errGov("empty task prompt")
	}
	return nil
}

type govErr struct{ msg string }

func (e govErr) Error() string { return "governance abort: " + e.msg }
func (e govErr) Unwrap() error { return ErrGovernanceAbort }

func errGov(msg string) error { return govErr{msg} }

// violation halts the trajectory on a policy violation (runaway retries).
func violation(t *Trace, cfg Config) bool {
	return t.Retries > cfg.MaxRetries
}
